Retry on API request timeout

## Background

Currently when the connection between where `vsce` is running and the Marketplace API times out, the publishing fails. This becomes especially problematic if packages for multiple platforms are being published at the same time, e.g.

```
vsce publish --pre-release --no-git-tag-version --packagePath $(find . -iname *.vsix)
```

(where `*.vsix` can match 8 files)

We ran into this recently in https://github.com/hashicorp/vscode-terraform/actions/runs/7540203435/job/20524327288#step:3:22

```
Publishing 'hashicorp.terraform (darwin-x64) v2.29.2024011611'...
Extension URL (might take a few minutes): https://marketplace.visualstudio.com/items?itemName=hashicorp.terraform
Hub URL: https://marketplace.visualstudio.com/manage/publishers/hashicorp/extensions/terraform/hub
Published hashicorp.terraform (darwin-x64) v2.29.2024011611.
Publishing 'hashicorp.terraform (web) v2.29.2024011611'...
Extension URL (might take a few minutes): https://marketplace.visualstudio.com/items?itemName=hashicorp.terraform
Hub URL: https://marketplace.visualstudio.com/manage/publishers/hashicorp/extensions/terraform/hub
Published hashicorp.terraform (web) v2.29.2024011611.
Publishing 'hashicorp.terraform (linux-arm64) v2.29.2024011611'...
Extension URL (might take a few minutes): https://marketplace.visualstudio.com/items?itemName=hashicorp.terraform
Hub URL: https://marketplace.visualstudio.com/manage/publishers/hashicorp/extensions/terraform/hub
Published hashicorp.terraform (linux-arm64) v2.29.2024011611.
Publishing 'hashicorp.terraform (linux-x64) v2.29.2024011611'...
Error: Request timeout: /_apis/gallery/publishers/hashicorp/extensions/terraform
Error: Process completed with exit code 1.
```

It leaves the extension in a half-published state, which is not ideal.

The easiest fix is to run this again with `--skip-duplicates` but it would be great if there was no need for such retry in the first place.

## Proposal

Have vsce retry on API failures like timeout automatically a few times with some backoff in between.

e.g. retry 3 times with ~5-10 seconds in between?

--- a/package.json
+++ b/package.json
@@ -35,7 +35,7 @@
 		"watch:test": "npm run test -- --watch"
 	},
 	"engines": {
-		"node": ">= 14"
+		"node": ">= 16"
 	},
 	"dependencies": {
 		"azure-devops-node-api": "^12.5.0",
@@ -70,7 +70,7 @@
 		"@types/mime": "^1",
 		"@types/minimatch": "^3.0.3",
 		"@types/mocha": "^7.0.2",
-		"@types/node": "^14.17.32",
+		"@types/node": "^16.11.7",
 		"@types/read": "^0.0.28",
 		"@types/semver": "^6.0.0",
 		"@types/tmp": "^0.2.2",

--- a/package.json
+++ b/package.json
@@ -41,6 +41,7 @@
 		"azure-devops-node-api": "^12.5.0",
 		"chalk": "^2.4.2",
 		"cheerio": "^1.0.0-rc.9",
+		"cockatiel": "^3.1.2",
 		"commander": "^6.2.1",
 		"form-data": "^4.0.0",
 		"glob": "^7.0.6",

--- a/src/publish.ts
+++ b/src/publish.ts
@@ -12,6 +12,7 @@ import { validatePublisher } from './validation';
 import { GalleryApi } from 'azure-devops-node-api/GalleryApi';
 import FormData from 'form-data';
 import { basename } from 'path';
+import { IterableBackoff, handleWhen, retry } from 'cockatiel';
 
 const tmpName = promisify(tmp.tmpName);
 
@@ -262,7 +263,14 @@ async function _publishSignedPackage(api: GalleryApi, packageName: string, packa
 		header: `--${form.getBoundary()}${lineBreak}Content-Disposition: attachment; name=sigzip; filename=${sigzipName}${lineBreak}Content-Type: application/octet-stream${lineBreak}${lineBreak}`
 	});
 
-	return await api.publishExtensionWithPublisherSignature(undefined, form, manifest.publisher, manifest.name, extensionType);
+	const publishWithRetry = retry(handleWhen(err => err.message.includes('timeout')), {
+		maxAttempts: 3,
+		backoff: new IterableBackoff([5_000, 10_000, 20_000])
+	});
+
+	return await publishWithRetry.execute(async () => {
+		return await api.publishExtensionWithPublisherSignature(undefined, form, manifest.publisher, manifest.name, extensionType);
+	});
 }
 
 export interface IUnpublishOptions extends IPublishOptions {

